spy_linux.go

```diff
@@ -32,17 +32,24 @@ func (c *pnConnIter) Next() *Connection {
 
 // cbConnections sets Connections()
 var cbConnections = func(processes bool) (ConnIter, error) {
+	// We read /proc/<pid>/net/tcp once per netns
+	netns := map[uint64]struct{}{}
 	buf := bufPool.Get().(*bytes.Buffer)
 	buf.Reset()
-	readFile(procRoot+"/net/tcp", buf)
-	readFile(procRoot+"/net/tcp6", buf)
+
 	var procs map[uint64]Proc
 	if processes {
 		var err error
-		if procs, err = walkProcPid(); err != nil {
+		if procs, err = walkProcPid(&netns, buf); err != nil {
 			return nil, err
 		}
 	}
+
+	if len(netns) == 0 {
```

Comment:
You're only using netns in this outer scope to store one bit of information. What do you think about returning an extra bool from walkProcPid?

```
bool found
if processes {
    var err error
    if procs, found, err = walkProcPid(buf); err != nil {
        return nil, err
    }
}

if !found {
    readFile(...)
    readFile(...)
}
```

Or, better yet, can we rely on readFile(...) always reading more than 0 bytes of data? In which case, it can be simplified even further...

```go
if processes {
    var err error
    if procs, err = walkProcPid(buf); err != nil {
        return nil, err
    }
}

if buf.Len() <= 0 {
    readFile(...)
    readFile(...)
}
```